<a href="https://colab.research.google.com/github/yejipark0514/Yeji-Park/blob/main/Plagiarism_Detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Final project- Plagarism Detection**


by Team **A9**: Namkyu Lee (nol5129), Tsung Ting Lee (tvl5472), Yeji Park (ykp5105),
Yearin Kim(kzy5152), Yu-Cheng Cheng (ykc5277)


**Background**

Plagiarism, often defined as a discrete offense, is using someone else’s work without attribution. In academic writing, plagiarism involves using words, ideas or information from a source without citing it correctly. Easy access to the Web and large databases has exacerbated plagiarism into a serious issue for educational institutions and students. Some plagiarism methods include copy-paste, paraphrasing, no proper use of quotation marks and idea and code plagiarism (Maurer et al., 2006). Paraphrasing is plagiarism if one lifes a direct phrase from another's work and changes a few words and if one does not properly credit the original author.


**Project Objective**

There are many plagiarism checkers existing in public but reliability and validity of these tools raise serious doubts. It is possible that the checkers might fail at determining legitimate plagiarism, may report false positives for common phrases and may provide different results because each website uses different techniques. On account of that, we would like to use Jaccard similarity, Cosine similarity, Levenshtein Similarity (Edit Distance) and N-gram for heatmap to see if the paraphrased text can pass the similartiy tests.



## Package install

In [ ]:
!pip install Levenshtein

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 174 kB 4.9 MB/s 
     |████████████████████████████████| 2.2 MB 18.9 MB/s 


In [ ]:
!pip install fuzzywuzzy

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
!pip install nltk

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


## Input file

In [ ]:
import re
train_text = "Org_file.txt"
test_text = "rephrased.txt"
unrelated_text = "unrelated.txt"
editPad = "editpad.txt"
paraphraser = "paraphraser_io.txt"
spinBot = "Spinbot.txt"
paraphrasing_tool = "paraphrasing_tool_com.txt"
prepostseo = "prepostseo.txt"

# read training data

with open(train_text) as f:
  train_text = f.read().lower()

# apply preprocessing (remove text inside square and curly brackets and rem punc)
train_text = re.sub(r"\[.*\]|\{.*\}", "", train_text)
train_text = re.sub(r'[^\w\s]', "", train_text)

with open(test_text) as r:
    test_text = r.read().lower()

# apply preprocessing (remove text inside square and curly brackets and rem punc)
test_text = re.sub(r"\[.*\]|\{.*\}", "", test_text)
test_text = re.sub(r'[^\w\s]', "", test_text)

with open(unrelated_text) as r:
    unrelated_text = r.read().lower()

# apply preprocessing (remove text inside square and curly brackets and rem punc)
unrelated_text = re.sub(r"\[.*\]|\{.*\}", "", unrelated_text)
unrelated_text = re.sub(r'[^\w\s]', "", unrelated_text)

with open(editPad) as r:
    editPad = r.read().lower()

# apply preprocessing (remove text inside square and curly brackets and rem punc)
editPad = re.sub(r"\[.*\]|\{.*\}", "", editPad)
editPad = re.sub(r'[^\w\s]', "", editPad)

with open(paraphraser) as r:
    paraphraser = r.read().lower()

# apply preprocessing (remove text inside square and curly brackets and rem punc)
paraphraser = re.sub(r"\[.*\]|\{.*\}", "", paraphraser)
paraphraser = re.sub(r'[^\w\s]', "", paraphraser)

with open(spinBot) as r:
    spinBot = r.read().lower()

# apply preprocessing (remove text inside square and curly brackets and rem punc)
spinBot = re.sub(r"\[.*\]|\{.*\}", "", spinBot)
spinBot = re.sub(r'[^\w\s]', "", spinBot)

with open(paraphrasing_tool) as r:
    paraphrasing_tool = r.read().lower()

# apply preprocessing (remove text inside square and curly brackets and rem punc)
paraphrasing_tool = re.sub(r"\[.*\]|\{.*\}", "", paraphrasing_tool)
paraphrasing_tool = re.sub(r'[^\w\s]', "", paraphrasing_tool)

with open(prepostseo) as r:
    prepostseo = r.read().lower()

# apply preprocessing (remove text inside square and curly brackets and rem punc)
prepostseo = re.sub(r"\[.*\]|\{.*\}", "", prepostseo)
prepostseo = re.sub(r'[^\w\s]', "", prepostseo)

#tokenize the text data
train_text_token = train_text.split()
test_text_token = test_text.split()
unrelated_text_token = unrelated_text.split()
editPad_token = editPad.split()
paraphraser_token = paraphraser.split()
spinBot_token = spinBot.split()
paraphrasing_tool_token = paraphrasing_tool.split()
prepostseo_token = prepostseo.split()


import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from numpy import dot
from numpy.linalg import norm


# sw contains the list of stopwords
sw = stopwords.words('english')
l1 = [];l2 = [];l3 = [];l4 = [];l5 = [];l6 = [];l7 = [];l8 = [];l9 = [];l10 = [];l11 = [];l12 = [];l13 = [];l14 = [];

# remove stop words from the string
X_set = {w for w in train_text_token if not w in sw}
Y_set = {w for w in test_text_token if not w in sw}
Z_set = {w for w in unrelated_text_token if not w in sw}
A_set = {w for w in editPad_token if not w in sw}
B_set = {w for w in paraphraser_token if not w in sw}
C_set = {w for w in spinBot_token if not w in sw}
D_set = {w for w in paraphrasing_tool_token if not w in sw}
E_set = {w for w in prepostseo_token if not w in sw}

print(len(A_set))
print("Stopword removed:", A_set)


364
Stopword removed: {'parallel', 'claim', 'lots', 'tasks', 'bysshe', 'returned', 'language', 'style', 'analyzer', 'really', 'exercise', 'byrons', 'work', 'summer', 'shared', 'romantic', 'analysis', 'implement', 'although', 'wonder', 'discouraged', 'meeting', 'believed', 'college', 'professor', 'boolean', 'figuresalan', 'never', 'turing', 'etc', 'much', 'improve', 'trapped', 'mit', 'teaches', 'turned', 'ever', 'formal', 'defense', 'telephone', 'close', 'electronic', 'related', 'third', 'equations', 'became', 'left', 'calculations', 'combined', 'one', 'used', 'basic', 'asking', 'knew', 'able', 'due', 'probability', 'nature', 'computable', 'made', 'called', 'scientific', 'introduced', 'shannonwe', 'earlier', 'limited', 'poetic', 'stretched', 'open', 'electromagnetic', 'programming', 'frankenstein', 'lord', 'wheels', 'charles', 'michigan', 'uncertainty', 'childhood', 'developed', 'physics', 'electrical', 'uncountable', 'cambridge', 'humanoid', 'read', 'phase', 'many', 'logical', 'shannon

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


#**Methods**
## Jaccard Similarity, Cosine similarity, and Edit Distance


#### **Jaccard Similarity**
In this part, we use Jaccard similarity by calculating the intersection of two documents over the union of two documents.

We compare the scores between the unrelated text and 6 rephrased texts.


In [ ]:
#define Jaccard Similarity function
def jaccard(list1, list2):
    intersection = len(list(set(list1).intersection(list2)))
    union = (len(list1) + len(list2)) - intersection
    return float(intersection) / union

#find Jaccard Similarity between the two sets (original and unrelated)
jaccard(X_set, Z_set)

NameError: ignored

In [ ]:
setList = [("Quillbot",X_set, Y_set),
           ("editPad", X_set, A_set),
           ("paraphraser",X_set,B_set),
           ("spinBot", X_set, C_set),
           ("paraphrasing_tool", X_set,D_set),
           ("prepostseo", X_set,E_set)]
for x, y, z in setList:
  l = jaccard(y, z)
  print(x, " - ", l)

Quillbot  -  0.43796992481203006
editPad  -  0.5687103594080338
paraphraser  -  0.504149377593361
spinBot  -  0.2644230769230769
paraphrasing_tool  -  0.5417515274949084
prepostseo  -  0.4516728624535316


#### **Cosine Similarity**
In this part, we use cosine similarity. We transformed two input sets into 0 and 1. Then, we calculated the dot product of two vectors over the product of two normalized vectors.

In [ ]:
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from numpy import dot
from numpy.linalg import norm


l1 = [];l2 = [];l3 = [];l4 = [];l5 = [];l6 = [];l7 = [];l8 = [];l9 = [];l10 = [];l11 = [];l12 = [];l13 = [];l14 = [];
# Use bag of word to create vector
rvector = X_set.union(Y_set)
for w in rvector:
    if w in X_set: l1.append(1) # create a vector
    else: l1.append(0)
    if w in Y_set: l2.append(1)
    else: l2.append(0)


urvector = X_set.union(Z_set)
for w in urvector:
    if w in X_set: l3.append(1) # create a vector
    else: l3.append(0)
    if w in Z_set: l4.append(1)
    else: l4.append(0)

rvector2 = X_set.union(A_set)
for w in rvector2:
    if w in X_set: l5.append(1) # create a vector
    else: l5.append(0)
    if w in Y_set: l6.append(1)
    else: l6.append(0)

rvector3 = X_set.union(B_set)
for w in rvector3:
    if w in X_set: l7.append(1) # create a vector
    else: l7.append(0)
    if w in Y_set: l8.append(1)
    else: l8.append(0)

rvector4 = X_set.union(C_set)
for w in rvector4:
    if w in X_set: l9.append(1) # create a vector
    else: l9.append(0)
    if w in Y_set: l10.append(1)
    else: l10.append(0)

rvector5 = X_set.union(D_set)
for w in rvector5:
    if w in X_set: l11.append(1) # create a vector
    else: l11.append(0)
    if w in Y_set: l12.append(1)
    else: l12.append(0)

rvector6 = X_set.union(E_set)
for w in rvector6:
    if w in X_set: l13.append(1) # create a vector
    else: l13.append(0)
    if w in Y_set: l14.append(1)
    else: l14.append(0)
# calculate the cosine similarity
rcosine = dot(l1, l2)/(norm(l1)*norm(l2))
urcosine = dot(l3, l4)/(norm(l3)*norm(l4))
rcosine2 = dot(l5, l6)/(norm(l5)*norm(l6))
rcosine3 = dot(l7, l8)/(norm(l7)*norm(l8))
rcosine4 = dot(l9, l10)/(norm(l9)*norm(l10))
rcosine5 = dot(l11, l12)/(norm(l11)*norm(l12))
rcosine6 = dot(l13, l14)/(norm(l13)*norm(l14))

file_list = [("Quillbot",rcosine),
             ("editPad", rcosine2),
             ("paraphraser",rcosine3),
             ("spinBot", rcosine4),
             ("paraphrasing_tool", rcosine5),
             ("prepostseo", rcosine6)]

for x, y in file_list:
  # print("cosine similarity for paraphrasing using ", x, " is:", y)
  print(x, " - ", y)
print("cosine similarity for unrelated essay: ", urcosine)

Quillbot  -  0.6091924869058762
editPad  -  0.7432301981157167
paraphraser  -  0.7418050190515105
spinBot  -  0.7461053740282597
paraphrasing_tool  -  0.7266504872319705
prepostseo  -  0.7534418865249852
cosine similarity for unrelated essay:  0.08744931082816979


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


#### **Edit Distance similarity ratio**
Edit distance is another way to detect the similarity between texts. In this package, we use token_set_ratio function. It does not punish the sorting, and it takes out the common tokens before calculate the edit distance. The function return a score from 0 to 100. The higher the score is, the more similar the two texts are.

In [ ]:
import fuzzywuzzy
from fuzzywuzzy import fuzz
from fuzzywuzzy import process

# similarity score for original text and rephrased text
ratio = fuzz.token_set_ratio(X_set, Y_set)
print('Similarity score: {}'.format(ratio))

Similarity score: 79


In [ ]:
# similarity score for original text and unrelated text
ratio = fuzz.token_set_ratio(X_set, Z_set)
print('Similarity score: {}'.format(ratio))

Similarity score: 45


In [ ]:
for x, y, z in setList:
  l = fuzz.token_set_ratio(y, z)
  print(x, " - ", l)

Quillbot  -  79
editPad  -  85
paraphraser  -  83
spinBot  -  69
paraphrasing_tool  -  85
prepostseo  -  81




# **Visualization**
NLP (nltk) & N-gram (heatmap)

In [ ]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [ ]:
import re
from nltk.util import ngrams, pad_sequence, everygrams
from nltk.tokenize import word_tokenize
from nltk.lm import MLE, WittenBellInterpolated
import numpy as np
import plotly.graph_objects as go
from scipy.ndimage import gaussian_filter
import statistics as st

## **Train file**

In [ ]:
# Choose the n-gram number
n = 4

# Pad the text and tokenize
training_data = list(pad_sequence(word_tokenize(train_text),
                                  n,pad_left=True,left_pad_symbol="<s>"))

# Generate ngram based on ngram number
ngrams = list(everygrams(training_data, max_len=n))
print("Number of ngrams:", len(ngrams))

# Build N-gram Language Model
model = WittenBellInterpolated(n)
model.fit([ngrams], vocabulary_text=training_data)
print(model.vocab)

Number of ngrams: 4154
<Vocabulary with cutoff=1 unk_label='<UNK>' and 444 items>


## **Test file**


In [ ]:
# Tokenize test file
testing_data = list(pad_sequence(word_tokenize(test_text), n,
                                 pad_left=True,
                                 left_pad_symbol="<s>"))
print("Length of test data:", len(testing_data))

Length of test data: 1010


In [ ]:
# Assign Scores
scores = []
for i, item in enumerate(testing_data[n-1:]):
    b = model.score(item, testing_data[i:i+n-1])
    scores.append(b)

scores_np = np.array(scores)

# Set Width and Height
width = 8
height = np.ceil(len(testing_data)/width).astype("int32")
print("Width, Height:", width, ",", height)

# Copy scores to rectangular blank array
z = np.zeros(width*height)
z[:len(scores_np)] = scores_np
diff = len(z) - len(scores_np)

# Apply gaussian smoothing for aesthetics
z = gaussian_filter(z, sigma=1.0)

# Reshape to fit rectangle
z = z.reshape(-1, width)

# Format labels
labels = [" ".join(testing_data[i:i+width]) for i in range(n-1, len(testing_data), width)]
labels_individual = [x.split() for x in labels]
labels_individual[-1] += [""]*diff
labels = [f"{x:60.60}" for x in labels]

print(st.mean(scores))

Width, Height: 8 , 127
0.1870091792774967


### Heatmap

In [ ]:
# Create heatmap for visualization
fig = go.Figure(data=go.Heatmap(
                z=z, x0=0, dx=1,
                y=labels, zmin=0, zmax=1,
                customdata=labels_individual,
                hovertemplate='%{customdata} <br><b>Score:%{z:.3f}<extra></extra>',
                colorscale="burg"))
fig.update_layout({"height":height*28, "width":1000, "font":{"family":"Courier New"}})
fig['layout']['yaxis']['autorange'] = "reversed"
fig.show()

In the heatmap, we can distinguish the similarity part in the text. The darker the color in the heatmap, the more similar the texts are. It is a useful visualization tool to check the similarity part and corrrect the indicated part to prevent plagiarism.

### Results

**Unrelated text results:**
*   Jaccard similarity: 0.043
*   Cosine similarity: 0.087
*   Edit Distance: 45
---------------------
**Paraphrasing tools**

**Quillbot:**
*   Jaccard similarity: 0.438
*   Cosine similarity: 0.609
*   Edit Distance: 79

**editPad:**
*   Jaccard similarity:  0.569
*   Cosine similarity: 0.743
*   Edit Distance: 85

**paraphraser:**
*   Jaccard similarity: 0.504
*   Cosine similarity: 0.742
*   Edit Distance: 83

**spinBot:**
*   Jaccard similarity: 0.264
*   Cosine similarity: 0.746
*   Edit Distance: 69

**paraphrasing_tool:**
*   Jaccard similarity: 0.542
*   Cosine similarity: 0.727
*   Edit Distance: 85

**prepostseo:**
*   Jaccard similarity: 0.452
*   Cosine similarity: 0.753
*   Edit Distance: 81


### Conclusion
The results showed that text with paraphrasing has a much higher similarity score in all similarity measures. Unrelated text will only have a small value of similarity score. We can conclude that both Jaccard and Cosine similarity measurements are suitable for plagiarism checking. Each similarity measurement has its strength and weakness. When considering duplicate terms in the document, Jaccard similarity will only count the term once. However, Cosine similarity will consider duplicates. In future work, we could use different similarity measurements depending on different situations. We also believe that Edit distance is a good measurement for plagiarism. Nowadays, people plagiarize other's work by inserting, deleting, and substituting words, and edit distance calculate score for it. Some researchers also study using edit distance for a plagiarism checking system and produce good results.

To sum up, there are more similarity measurements for plagiarism detection. In this work, we prove that by only using the three most basic similarity measurements, we can distinguish paraphrasing text created by different paraphrasing websites.

Works Cited

Maurer, Hermann & Kappe, Frank & Zaka, Bilal. (2006). Plagiarism - A Survey. Journal of Universal Computer Science. 12. 1050-1084.
https://www.researchgate.net/publication/220017646_Plagiarism_-_A_Survey/link/0912f514817642d96e000000/download
